pkg/apis/discovery/validation/validation_test.go

```diff
@@ -313,7 +328,37 @@ func TestValidateEndpointSlice(t *testing.T) {
 					Protocol: protocolPtr(api.ProtocolTCP),
 				}},
 				Endpoints: []discovery.Endpoint{{
-					Addresses: generateAddresses(1),
+					Addresses: generateIPAddresses(1),
+					Hostname:  utilpointer.StringPtr("valid-123"),
+				}},
+			},
+		},
+		"bad-ip": {
+			expectedErrors: 1,
+			endpointSlice: &discovery.EndpointSlice{
+				ObjectMeta:  standardMeta,
+				AddressType: addressTypePtr(discovery.AddressTypeIP),
+				Ports: []discovery.EndpointPort{{
+					Name:     utilpointer.StringPtr("http"),
+					Protocol: protocolPtr(api.ProtocolTCP),
+				}},
+				Endpoints: []discovery.Endpoint{{
+					Addresses: []string{"123.456.789.012"},
+					Hostname:  utilpointer.StringPtr("valid-123"),
+				}},
+			},
+		},
+		"bad-fqdn": {
```

Comment:
Can we get some other test cases? E.g. 
* IP addresses
* including parts of a URL - http://foo.example.com, lavalamp@foo.example.com, foo.example.com/mypage, foo.example.com#myanchor

I assume all of these should be illegal?

Reply:
I've added some more thorough validation that follows the pattern of other apimachinery validation tests, along with some small additions to the corresponding EndpointSlice validation tests. Amusingly, IPv4 addresses actually pass this validation, and that actually seems reasonable. Since this is validating that there are at least 3 DNS labels separated by `.`, and DNS labels can contain purely numbers, an IP address is considered valid. Though of course it would be interpreted as a domain name here. A common example of a numeric domain is 911.gov, though of course the TLD is not numeric, I don't think we can safely assume that all valid FQDNs will never have a numeric TLD.

Reply:
> Since this is validating that there are at least 3 DNS labels separated by .

Is that the behavior you want? `example.com` isn't valid?

Reply:
It seems like it's relatively difficult to nail down a precise spec for FQDNs. Most examples do include 3 segments, including all examples found in relevant RFCs that I can find, but there's no evidence that it's actually necessary. It's entirely possible that `example.com` could be a FQDN in certain use cases. Given that along with the current `IsFullyQualifiedName` function failing on a FQDN with a trailing `.`, it seemed worthwhile to create a new validation function here. I've added a `IsFullyQualifiedDomainName` right below `IsFullyQualifiedName` and added comments to differentiate the validations used here.